File: emisor-python/algorithms/crc32.py
CRC32_POLYNOMIAL = "100000100110000010001110110110111"
CRC_SIZE = 32


def calculate_remainder(bits):
    poly_len = len(CRC32_POLYNOMIAL)
    dividend = list(bits + "0" * CRC_SIZE)

    for i in range(len(bits)):
        if dividend[i] == "1":
            for j in range(poly_len):
                dividend[i + j] = str(
                    int(dividend[i + j]) ^ int(CRC32_POLYNOMIAL[j])
                )

    remainder = "".join(dividend[len(bits):])
    return remainder


def calculate_integrity(bits):
    crc = calculate_remainder(bits)
    return bits + crc


def verify_integrity(frame):
    if len(frame) <= CRC_SIZE:
        raise ValueError("The frame is too short to contain a valid CRC-32")

    data = frame[:-CRC_SIZE]
    received_crc = frame[-CRC_SIZE:]
    calculated_crc = calculate_remainder(data)

    error = calculated_crc != received_crc
    return data, error


def correct_message(frame):
    data, error = verify_integrity(frame)
    return data, error

File: receptor-go/algorithms/crc32.go
package algorithms

import "fmt"

const CRC32Polynomial = "100000100110000010001110110110111"
const CRCSize = 32

func calculateRemainder(bits string) string {
	polyLen := len(CRC32Polynomial)
	dividend := make([]byte, len(bits)+CRCSize)
	copy(dividend, bits)

	for i := len(bits); i < len(dividend); i++ {
		dividend[i] = '0'
	}

	for i := 0; i < len(bits); i++ {
		if dividend[i] == '1' {
			for j := range polyLen {
				dividend[i+j] = xorBit(dividend[i+j], CRC32Polynomial[j])
			}
		}
	}

	return string(dividend[len(bits):])
}

func xorBit(a, b byte) byte {
	if a == b {
		return '0'
	}
	return '1'
}

func VerifyIntegrity(frame string) (string, bool, error) {
	if len(frame) <= CRCSize {
		return "", false, fmt.Errorf("the frame is too short to contain a valid CRC-32")
	}

	data := frame[:len(frame)-CRCSize]
	receivedCRC := frame[len(frame)-CRCSize:]
	calculatedCRC := calculateRemainder(data)

	hasError := calculatedCRC != receivedCRC

	return data, hasError, nil
}
